### PR TITLE
[neighbor_advertiser] Verify that DIPs returned from ferret are not in device VLAN

### PR DESCRIPTION
**Pull request by daall**
**- What I did**
I added a check to neighbor_advertiser to make sure that the DIPs returned from ferret are not in any of the device's VLANs.

**- How I did it**
- I added a helper method to query CONFIG_DB for VLAN interfaces and check that a given IP does not fall in that range.
- I used the helper method to check that DIPs returned from ferret did not fall in any of the device's VLANs.
- I added a retry to attempt to get other DIPs from ferret before moving on to the next ferret VIP.

**- How to verify it**
1. Run the script against a response object that contains an IPv4 address that falls inside one of the DUTs VLANs. The script should retry 3 times and then fail.
2. Run the script against a response object that contains an IPv4 address that does not fall inside of the DUTs VLANs. The script should succeed.

Signed-off-by: Danny Allen (daall@microsoft.com)

### CODE REVIEW

**Comment by daall** [comment]
retest this please

File: scripts/neighbor_advertiser [hunk 1]
```diff
@@ -308,24 +331,30 @@ def post_neighbor_advertiser_slice(ferret_service_vip):
...
-    else:
-        log_error('Failed to set up neighbor advertiser slice, vip: {}, no response obtained'.format(ferret_service_vip))
 
-    return ferret_server_ipv4_addr
+        if not is_dip_in_device_vlan(ferret_server_ipv4_addr):
+            log_info('Successfully set up neighbor advertiser slice, vip: {}, dip: {}'.format(ferret_service_vip, ferret_server_ipv4_addr))
+            return ferret_server_ipv4_addr
+
+    log_error('Failed to set up neighbor advertiser slice, vip: {}, returned dips were in device VLAN'.format(ferret_service_vip))
+    return None
```
Comment (yxieca) [inline scripts/neighbor_advertiser]: Not a major issue.

But I suggest you to follow the convention above, put failure handling code in if block and the main return becomes successful return.

**Comment by daall** [comment]
retest this please

File: scripts/neighbor_advertiser [hunk 2]
```diff
@@ -308,23 +328,34 @@ def post_neighbor_advertiser_slice(ferret_service_vip):
...
-        ferret_server_ipv4_addr = neighbor_advertiser_configuration['ipv4Addr']
-        log_info('Successfully set up neighbor advertiser slice, vip: {}, dip: {}'.format(ferret_service_vip, ferret_server_ipv4_addr))
-    elif response:
-        log_error('Failed to set up neighbor advertiser slice, vip: {}, error_code: {}, error_content: {}'.format(ferret_service_vip, response.status_code, response.content))
-    else:
-        log_error('Failed to set up neighbor advertiser slice, vip: {}, no response obtained'.format(ferret_service_vip))
+        ferret_dip = neighbor_advertiser_configuration['ipv4Addr']
+
+        if not is_dip_in_device_vlan(ferret_dip):
+            save_as_json(neighbor_advertiser_configuration, NEIGHBOR_ADVERTISER_RESPONSE_CONFIG_PATH)
```
Comment (yxieca) [inline scripts/neighbor_advertiser]: I think you want to generate a log here when "is_dip_in_device_vlan(ferret_dip)" is true and the main branch can log success and return directly?

File: scripts/neighbor_advertiser [hunk 3]
```diff
@@ -82,16 +84,34 @@ def log_error(msg):
...
+is_dip_in_device_vlan = None
+
+
+def get_device_vlan_checker():
+    global is_dip_in_device_vlan
+
+    vlan_interfaces = config_db.get_table('VLAN_INTERFACE')
+
+    def check_device(ferret_server_ip_addr):
+        return any(IPAddress(ferret_server_ip_addr) in IPNetwork(vlan_int[1]) for vlan_int in vlan_interfaces.iterkeys())
```
Comment (lguohan) [inline scripts/neighbor_advertiser]: >vlan_int[1]) for vlan_int in vlan_interfaces.iterkeys [](start = 65, length = 53)

what if vlan_interface has ipv6 address? we should only filter ipv4 address here.

Comment (lguohan) [reply scripts/neighbor_advertiser]: we also have a case where vlan interface only contains interface name, but not IP address. we should check vlan_int[1] is available or not. you can sync with prince on this case.

---
In reply to: [328399686](https://github.com/Azure/sonic-utilities/pull/670#discussion_r328399686) [](ancestors = 328399686)

Comment (yxieca) [reply scripts/neighbor_advertiser]: Good catch!

File: scripts/neighbor_advertiser [hunk 4]
```diff
@@ -308,24 +328,36 @@ def post_neighbor_advertiser_slice(ferret_service_vip):
...
+            log_error('Failed to set up neighbor advertiser slice, vip: {}, error_code: {}, error_content: {}'.format(ferret_service_vip, response.status_code, response.content))
+            return None
 
-    if response and response.ok:
         neighbor_advertiser_configuration = json.loads(response.content)
-        save_as_json(neighbor_advertiser_configuration, NEIGHBOR_ADVERTISER_RESPONSE_CONFIG_PATH)
         ferret_server_ipv4_addr = neighbor_advertiser_configuration['ipv4Addr']
+
+        # Retry the request if the provided DIP is in the device VLAN
+        if is_dip_in_device_vlan(ferret_server_ipv4_addr):
```
Comment (prsunny) [inline scripts/neighbor_advertiser]: So this is the global variable assigned to a function name. Why don't we define this as a function itself for better clarity?

Comment (prsunny) [reply scripts/neighbor_advertiser]: In that case, a class would be ideal. But another (simple) option is to make the Vlan interface global and fetch it in main. What do you think?

File: scripts/neighbor_advertiser [hunk 5]
```diff
@@ -82,16 +84,50 @@ def log_error(msg):
...
+    global vlan_interface_query
+
+    # Lazy load the vlan interfaces the first time we run this check.
+    if not vlan_interface_query:
+        vlan_interface_query = config_db.get_table('VLAN_INTERFACE')
+
+    ferret_dip = IPAddress(ferret_dip)
+
+    for vlan_interface in vlan_interface_query.iterkeys():
+        vlan_subnet = vlan_interface[1]
```
Comment (prsunny) [inline scripts/neighbor_advertiser]: As discussed, you can reuse this function for this check - `is_ip_prefix_in_key(self, key)`

**Comment by yxieca** [comment]
retest this please

**Pull request merged**